x-pack/plugins/observability_solution/investigate_app/server/services/get_entities.ts

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import { z } from '@kbn/zod';
+import { entityLatestSchema } from '@kbn/entities-schema';
+import { GetEntitiesResponse, EntityWithSource, EntitySource } from '@kbn/investigation-shared';
+import { QueryDslQueryContainer } from '@elastic/elasticsearch/lib/api/typesWithBodyKey';
+import { IndicesIndexState } from '@elastic/elasticsearch/lib/api/typesWithBodyKey';
+
+import { InvestigateAppRequestHandlerContext } from '../routes/types';
+import { EntitiesESClient } from '../clients/create_entities_es_client';
+import {
+  SERVICE_ENTITIES_LATEST_ALIAS,
+  CONTAINER_ENTITIES_LATEST_ALIAS,
+  HOST_ENTITIES_LATEST_ALIAS,
+} from '../clients/create_entities_es_client';
+
+// the official types do not explicitly define sourceIndex in the schema, but it is present in the data at the time of writing this
+type EntitiesLatest = z.infer<typeof entityLatestSchema> & { sourceIndex: string[] };
+
+export async function getEntitiesWithSource({
+  context,
+  serviceEnvironment,
+  serviceName,
+  containerId,
+  hostName,
+  entitiesEsClient,
+}: {
+  context: InvestigateAppRequestHandlerContext;
```

Comment:
If possible let's try to not leak route/request details into the services. Here we can replace the whole request handler context with the esClient, and do the wiring in the route handler.